[Pytorch Edge] Support profiling kineto events from external source

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64397

Summary:
This diff exposes a way to add events to kineto profiler from external
source.
This can be a backend that executes a subgraph and wants to record this
execution in kineto profiler.
This diff also adds "backend" metadata to identify the backend an event
would have executed on.

Test Plan:
test_lite_interpreter

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30710710](https://our.internmc.facebook.com/intern/diff/D30710710)

Here is a chrome trace output for example delegated backend
![image](https://user-images.githubusercontent.com/17488981/131749023-71062b4c-9d1f-428a-9bb1-6bd45b02d194.png)

kimishpatel committed:
[Pytorch Edge] Support profiling kineto events from external source

Summary:
This diff exposes a way to add events to kineto profiler from external
source.
This can be a backend that executes a subgraph and wants to record this
execution in kineto profiler.
This diff also adds "backend" metadata to identify the backend an event
would have executed on.

Test Plan:
test_lite_interpreter

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30710710).

kimishpatel committed:
Update on "[Pytorch Edge] Support profiling kineto events from external source"


Summary:
This diff exposes a way to add events to kineto profiler from external
source.
This can be a backend that executes a subgraph and wants to record this
execution in kineto profiler.
This diff also adds "backend" metadata to identify the backend an event
would have executed on.

Test Plan:
test_lite_interpreter

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30710710](https://our.internmc.facebook.com/intern/diff/D30710710)

Here is a chrome trace output for example delegated backend
![image](https://user-images.githubusercontent.com/17488981/131749023-71062b4c-9d1f-428a-9bb1-6bd45b02d194.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30710710).

raziel commented on torch/csrc/autograd/profiler_kineto.cpp:
```diff
@@ -458,6 +463,40 @@ std::string stacksToStr(const std::vector<std::string>& stacks, const char* deli
+  ctx_ptr->startThreadId = at::RecordFunction::currentThreadId();
+  ctx_ptr->debug_handle = debug_handle;
+
+  /* no support for input shapes now?
```
why's this?

kimishpatel replied:
Mainly to avoid overhead of shape profiling.

raziel reviewed (approved):
LGTM Thanks

Though somebody from Kineto should review the Kineto changes.

gdankel commented on torch/csrc/autograd/profiler_kineto.cpp:
```diff
@@ -61,13 +62,13 @@ struct KinetoThreadLocalState : public ProfilerThreadLocalState {
     }
-    std::string evt_name(fn.name().str());
-    auto end_time = getTimeUs();
+    auto end_time = ctx->endUS;
```
Can we assume that this is always set?

gdankel commented on torch/csrc/autograd/profiler_kineto.cpp:
```diff
@@ -389,7 +393,8 @@ void pushProfilingCallbacks(const std::unordered_set<at::RecordScope>& scopes) {
           }
 
-          state_ptr->reportClientActivity(fn, kineto_ctx_ptr);
+          kineto_ctx_ptr->endUS = getTimeUs();
```
Why move this here?

kimishpatel replied:
This is so that `reportClientActivity` can assume the start and end time are already encoded in ctx. When backend's report the runtime per op they will use indirectly use `reportClientActivity`. At the time of recording runtime of op, backends already have the duration/start and end time.

gdankel commented on torch/csrc/autograd/profiler_kineto.h:
```diff
@@ -333,6 +344,14 @@ struct TORCH_API ProfilerResult {
 #endif // USE_KINETO
 };
 
+TORCH_API void reportBackendEventToActiveKinetoProfiler(
```
Please add comment / documentation.

gdankel commented on torch/csrc/autograd/profiler_kineto.cpp:
```diff
@@ -458,6 +463,40 @@ std::string stacksToStr(const std::vector<std::string>& stacks, const char* deli
+  if (!state_ptr) {
+    return;
+  }
+  auto ctx_ptr = std::make_unique<KinetoObserverContext>();
```
While this is fine for the moment, I think we can optimize this a lot - it seems unnecessary to create the KinetoObserverContext in this case. In fact we can probably remove it from the profiler altogether...

kimishpatel replied:
I dont know if you meant for the backend event reporting only or generally in the profiler itself but I completely agree. I thought the same that we dont need the ObserverContext. My understanding was that the observer context was used by callback where enter callback construct one, push some info and return the context. Then record_function calls exit callback with the context which populates more information. For the purposes of the API here, it is not needed.

gdankel reviewed (approved):
I have some nits but in general this looks fine for now.
Definitely need to have a follow-up refactoring but that applies to the entire profiler.

kimishpatel committed:
Update on "[Pytorch Edge] Support profiling kineto events from external source"


Summary:
This diff exposes a way to add events to kineto profiler from external
source.
This can be a backend that executes a subgraph and wants to record this
execution in kineto profiler.
This diff also adds "backend" metadata to identify the backend an event
would have executed on.

Test Plan:
test_lite_interpreter

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30710710](https://our.internmc.facebook.com/intern/diff/D30710710)

Here is a chrome trace output for example delegated backend
![image](https://user-images.githubusercontent.com/17488981/131749023-71062b4c-9d1f-428a-9bb1-6bd45b02d194.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30710710).

kimishpatel committed:
Update on "[Pytorch Edge] Support profiling kineto events from external source"


Summary:
This diff exposes a way to add events to kineto profiler from external
source.
This can be a backend that executes a subgraph and wants to record this
execution in kineto profiler.
This diff also adds "backend" metadata to identify the backend an event
would have executed on.

Test Plan:
test_lite_interpreter

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30710710](https://our.internmc.facebook.com/intern/diff/D30710710)

Here is a chrome trace output for example delegated backend
![image](https://user-images.githubusercontent.com/17488981/131749023-71062b4c-9d1f-428a-9bb1-6bd45b02d194.png)


[ghstack-poisoned]

janeyx99 commented:
Reverting this PR as it failed the bazel test (PR signal was red as well) on HUD

kimishpatel commented:
> Reverting this PR as it failed the bazel test (PR signal was red as well) on HUD

Is the failing build trigerred on the PRs? I remember seeing all green for this PR's CI.
